### PR TITLE
[new release] passage (0.1.5)

### PR DESCRIPTION
**Pull request by thatportugueseguy**
Passage - used to store and manage access to shared secrets

- Project page: <a href="https://github.com/ahrefs/passage">https://github.com/ahrefs/passage</a>

##### CHANGES:

- Add replace-comment command
- Make new and edit commands usage uniform
- Remove bash completions opam install. Moved to debian package.

### CODE REVIEW

**Comment by raphael-proust** [comment]
CI errors:

some cram test errors on fedora-41:
```
#=== ERROR while compiling passage.0.1.5 ======================================#
# context              2.4.0~alpha1~dev | linux/x86_64 | ocaml-base-compiler.4.14.2 | pinned(https://github.com/ahrefs/passage/releases/download/0.1.5/passage-0.1.5.tbz)
# path                 ~/.opam/4.14/.opam-switch/build/passage.0.1.5
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p passage -j 255 @install @runtest
# exit-code            1
# env-file             ~/.opam/log/passage-7-c8a2aa.env
# output-file          ~/.opam/log/passage-7-c8a2aa.out
### output ###
# File "tests/replace_comment_command.t", line 1, characters 0-0:
# /usr/bin/git --no-pager diff --no-index --color=always -u _build/.sandbox/07755d2dc5d9f491400477623f2b4fde/default/tests/replace_comment_command.t _build/.sandbox/07755d2dc5d9f491400477623f2b4fde/default/tests/replace_comment_command.t.corrected
# diff --git a/_build/.sandbox/07755d2dc5d9f491400477623f2b4fde/default/tests/replace_comment_command.t b/_build/.sandbox/07755d2dc5d9f491400477623f2b4fde/default/tests/replace_comment_command.t.corrected
# index 064bf1b..07087bd 100644
# --- a/_build/.sandbox/07755d2dc5d9f491400477623f2b4fde/default/tests/replace_comment_command.t
# +++ b/_build/.sandbox/07755d2dc5d9f491400477623f2b4fde/default/tests/replace_comment_command.t.corrected
# @@ -29,34 +29,28 @@ Should succeed - replacing single-line comments with multiline comments
#    $ passage cat 00/secret1
#    (00/secret1) secret: single line
#    
# -  replaced again comments
# -  line 2
# +  replaced again comments\nline 2
#  
#  Should succeed - replacing multiline comments with multiline comments
#    $ echo "new comments\nline 2 of said new comments" | passage replace-comment 00/secret1
#    $ passage cat 00/secret1
#    (00/secret1) secret: single line
#    
# -  new comments
# -  line 2 of said new comments
# +  new comments\nline 2 of said new comments
#  
#  Should succeed - replacing multiline comments with multiline comments - in multiline secret
#    $ setup_multiline_secret_with_comments 00/secret2
#    $ echo "new comments\nline 2 of said new comments" | passage replace-comment 00/secret2
#    $ passage cat 00/secret2
#    
# -  new comments
# -  line 2 of said new comments
# +  new comments\nline 2 of said new comments
#    
#    (00/secret2) secret: line 1
#    (00/secret2) secret: line 2
#  
#  Should fail - comments with empty lines in the middle
#    $ echo "uno commento\n\ndos commentos" | passage replace-comment 00/secret1
# -  The comments are in an invalid format: secrets cannot have empty lines in the middle of the comments
# -  [1]
#    $ passage cat 00/secret1
#    (00/secret1) secret: single line
#    
# -  new comments
# -  line 2 of said new comments
# +  uno commento\n\ndos commentos
# File "tests/replace_command.t", line 1, characters 0-0:
# /usr/bin/git --no-pager diff --no-index --color=always -u _build/.sandbox/20d1e06d5ac8ee9e2243cbf046bcbd42/default/tests/replace_command.t _build/.sandbox/20d1e06d5ac8ee9e2243cbf046bcbd42/default/tests/replace_command.t.corrected
# diff --git a/_build/.sandbox/20d1e06d5ac8ee9e2243cbf046bcbd42/default/tests/replace_command.t b/_build/.sandbox/20d1e06d5ac8ee9e2243cbf046bcbd42/default/tests/replace_command.t.corrected
# index 6bedc16..ef01dcb 100644
# --- a/_build/.sandbox/20d1e06d5ac8ee9e2243cbf046bcbd42/default/tests/replace_command.t
# +++ b/_build/.sandbox/20d1e06d5ac8ee9e2243cbf046bcbd42/default/tests/replace_command.t.corrected
# @@ -214,7 +214,7 @@ Should succeed - correctly handle inputs without newline termination
#    $ printf "secret\n\nnew comments\nmultiple comments" | passage create 01/new_secret_singleline_with_comments_2
#    $ echo "wuuut\c" | passage replace 01/new_secret_singleline_with_comments_2
#    $ passage cat 01/new_secret_singleline_with_comments_2
# -  wuuut
# +  wuuut\c
#    
#    new comments
#    multiple comments
```

There seems to be some escaping problem. I find this a bit puzzling. Like how does the cram test end up with literal backslash-n sequences instead of newline character!?

For what it's worth, I have the same issue locally.

other cram test errors in extras (old opam):
```
#=== ERROR while compiling passage.0.1.5 ======================================#
# context              2.0.10 | linux/x86_64 | ocaml-base-compiler.4.14.2 | pinned(https://github.com/ahrefs/passage/releases/download/0.1.5/passage-0.1.5.tbz)
# path                 ~/.opam/4.14/.opam-switch/build/passage.0.1.5
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p passage -j 39 @install @runtest
# exit-code            1
# env-file             ~/.opam/log/passage-7-5f1023.env
# output-file          ~/.opam/log/passage-7-5f1023.out
### output ###
# File "tests/template_secrets_command.t", line 1, characters 0-0:
# /usr/bin/git --no-pager diff --no-index --color=always -u _build/.sandbox/3a9cdbbacfdb0b52ff3138a92758ad0f/default/tests/template_secrets_command.t _build/.sandbox/3a9cdbbacfdb0b52ff3138a92758ad0f/default/tests/template_secrets_command.t.corrected
# diff --git a/_build/.sandbox/3a9cdbbacfdb0b52ff3138a92758ad0f/default/tests/template_secrets_command.t b/_build/.sandbox/3a9cdbbacfdb0b52ff3138a92758ad0f/default/tests/template_secrets_command.t.corrected
# index 3a2c0c0..8a3b6f5 100644
# --- a/_build/.sandbox/3a9cdbbacfdb0b52ff3138a92758ad0f/default/tests/template_secrets_command.t
# +++ b/_build/.sandbox/3a9cdbbacfdb0b52ff3138a92758ad0f/default/tests/template_secrets_command.t.corrected
# @@ -13,6 +13,6 @@ Extract multiple secret names from a string
#  
#  Non-existent template file
#    $ passage template-secrets no_such_file
# -  Failed to parse the file : Sys_error("no_such_file: No such file or directory")
# +  Failed to parse the file : (Sys_error "no_such_file: No such file or directory")
#    [1]
#  
# File "tests/template_command.t", line 1, characters 0-0:
# /usr/bin/git --no-pager diff --no-index --color=always -u _build/.sandbox/f40f338b7a84aaa93521647459508865/default/tests/template_command.t _build/.sandbox/f40f338b7a84aaa93521647459508865/default/tests/template_command.t.corrected
# diff --git a/_build/.sandbox/f40f338b7a84aaa93521647459508865/default/tests/template_command.t b/_build/.sandbox/f40f338b7a84aaa93521647459508865/default/tests/template_command.t.corrected
# index bf39cce..1c83cbb 100644
# --- a/_build/.sandbox/f40f338b7a84aaa93521647459508865/default/tests/template_command.t
# +++ b/_build/.sandbox/f40f338b7a84aaa93521647459508865/default/tests/template_command.t.corrected
# @@ -58,7 +58,7 @@ Should succeed - template with multiple secrets
#  
#  Should fail - invalid template file
#    $ passage template non_existent_template $TARGET
# -  E: failed to substitute file : Sys_error("non_existent_template: No such file or directory")
# +  E: failed to substitute file : (Sys_error "non_existent_template: No such file or directory")
#    [1]
#  
#  Should fail - unable to decrypt a secret
# File "tests/edit_who_command.t", line 1, characters 0-0:
# /usr/bin/git --no-pager diff --no-index --color=always -u _build/.sandbox/0f672cc7d21f42bbcb678fe25d6f1610/default/tests/edit_who_command.t _build/.sandbox/0f672cc7d21f42bbcb678fe25d6f1610/default/tests/edit_who_command.t.corrected
# diff --git a/_build/.sandbox/0f672cc7d21f42bbcb678fe25d6f1610/default/tests/edit_who_command.t b/_build/.sandbox/0f672cc7d21f42bbcb678fe25d6f1610/default/tests/edit_who_command.t.corrected
# index b4b0d23..5ac9a3e 100644
# --- a/_build/.sandbox/0f672cc7d21f42bbcb678fe25d6f1610/default/tests/edit_who_command.t
# +++ b/_build/.sandbox/0f672cc7d21f42bbcb678fe25d6f1610/default/tests/edit_who_command.t.corrected
# @@ -12,7 +12,8 @@ Should succeed - edit recipients when the user is a recipient of the folder. Ref
#    $ PASSAGE_IDENTITY=robby.rob.key passage get 02/secret1
#    age: error: no identity matched any of the recipients
#    age: report unexpected or unhelpful errors at https://filippo.io/age/report
# -  E: failed to decrypt 02/secret1 : Failure("age --decrypt --identity $TESTCASE_ROOT/robby.rob.key : exit code 1")
# +  E: failed to decrypt 02/secret1 : (Failure
# +    "age --decrypt --identity $TESTCASE_ROOT/robby.rob.key : exit code 1")
#    [1]
#    $ passage who 02/secret1
#    bobby.bob
# @@ -55,7 +56,8 @@ Should succeed - use path to edit recipients instead of secret name
#    $ passage get 01/00/secret1
#    age: error: no identity matched any of the recipients
#    age: report unexpected or unhelpful errors at https://filippo.io/age/report
# -  E: failed to decrypt 01/00/secret1 : Failure("age --decrypt --identity $TESTCASE_ROOT/bobby.bob.key : exit code 1")
# +  E: failed to decrypt 01/00/secret1 : (Failure
# +    "age --decrypt --identity $TESTCASE_ROOT/bobby.bob.key : exit code 1")
#    [1]
#    $ passage who 01/00
#    poppy.pop
# @@ -82,7 +84,8 @@ Should succeed - refresh after edits for groups - should work as expected
#    $ passage get 03/secret1
#    age: error: no identity matched any of the recipients
#    age: report unexpected or unhelpful errors at https://filippo.io/age/report
# -  E: failed to decrypt 03/secret1 : Failure("age --decrypt --identity $TESTCASE_ROOT/bobby.bob.key : exit code 1")
# +  E: failed to decrypt 03/secret1 : (Failure
# +    "age --decrypt --identity $TESTCASE_ROOT/bobby.bob.key : exit code 1")
#    [1]
#    $ PASSAGE_IDENTITY=robby.rob.key passage get 03/secret1
#    (03/secret1) secret: single line
```

(goes on for a while longer)

Also somewhat puzzled by this. Although these seem to be mostly about the formatting (not the informational content) of error messages (not successful operations). So it's probably not an actual issue for end-users.

@thatportugueseguy Do you want to fix this upstream? Do you have something else in mind?


--------------

And then some unimportant stuff:

libevent error on macos:
```
#=== ERROR while compiling libevent.0.9.0 =====================================#
# context              2.3.0 | macos/arm64 | ocaml-base-compiler.5.2.1 | file:///Users/mac1000/opam-repository
# path                 ~/.opam/5.2.1/.opam-switch/build/libevent.0.9.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build make depend
# exit-code            2
# env-file             ~/.opam/log/libevent-26156-20fd50.env
# output-file          ~/.opam/log/libevent-26156-20fd50.out
### output ###
# Makefile:99: depend: No such file or directory
# gcc -I /Users/mac1000/.opam/5.2.1/lib/ocaml -MM *.c > depend
# event_stubs.c:13:10: fatal error: 'event.h' file not found
# #include <event.h>
#          ^~~~~~~~~
# 1 error generated.
# make: *** [depend] Error 1
```

re2 error on freebsd:
```

#=== ERROR while compiling re2.v0.16.0 ========================================#
# context              2.3.0~rc1 | freebsd/x86_64 | ocaml-base-compiler.5.2.0 | file:///home/opam/opam-repository
# path                 ~/.opam/5.2.0/.opam-switch/build/re2.v0.16.0
# command              ~/.opam/5.2.0/bin/dune build -p re2 -j 7
# exit-code            1
# env-file             ~/.opam/log/re2-84487-b105d9.env
# output-file          ~/.opam/log/re2-84487-b105d9.out
### output ###
# File "src/re2_c/dune", lines 5-26, characters 0-675:
#  5 | (rule (targets libre2_c_stubs.a dllre2_c_stubs.so)
#  6 |  (deps (source_tree libre2))
#  7 |  (action
# ....
# 24 |    \ncp libre2/obj/so/libre2.so dllre2_c_stubs.so\
# 25 |    \n%{make} -s -C libre2 clean\
# 26 |    \n")))
# (cd _build/default/src/re2_c && /usr/local/bin/bash -e -u -o pipefail -c '
# ARFLAGS=rsc
# CXXFLAGS="-Wall -O3 -g -pthread"
# if [ FreeBSD = `uname -s` ]; then
#      CXX=clang++
#      CXXFLAGS="$CXXFLAGS -I /usr/local/lib/re2"
# else
#      CXX=g++
# fi
# if ! true; then
# CXX="$CXX -m32"
# fi
# /usr/local/bin/gmake -s -C libre2 clean
# /usr/local/bin/gmake -s -C libre2 ARFLAGS="$ARFLAGS" CXX="$CXX" CXXFLAGS="$CXXFLAGS" obj/libre2.a obj/so/libre2.so
# cp libre2/obj/libre2.a libre2_c_stubs.a
# cp libre2/obj/so/libre2.so dllre2_c_stubs.so
# /usr/local/bin/gmake -s -C libre2 clean
# ')
# ld: error: version script assignment of 'global' to symbol '_ZlsRSoRKN3re211StringPieceE' failed: symbol not defined
# clang++: error: linker command failed with exit code 1 (use -v to see invocation)
# gmake: *** [Makefile:181: obj/so/libre2.so] Error 1
```

**Comment by thatportugueseguy** [comment]
>Also somewhat puzzled by this. Although these seem to be mostly about the formatting (not the informational content) of error messages (not successful operations). So it's probably not an actual issue for end-users.
>
>@thatportugueseguy Do you want to fix this upstream? Do you have something else in mind?

Again, strange outputs.. It's weird because the formatting change comes from the raised exceptions, not the code. They just bubble up until they are handled in a try/with and then printed out with `Printexc.to_string`..

**Comment by thatportugueseguy** [comment]
I don't think these "issues" will realistically affect anyone. Is it possible to merge as it is?

**Comment by raphael-proust** [comment]
I think some of the errors (on my local machine for sure) (and maybe possibly on fedora) is due to:

```
$ dash -c 'echo "foo\nbar" | grep foo'
foo
$ bash -c 'echo "foo\\nbar" | grep foo'
foo\nbar
```

and locally I have

```
$ file $(which sh)
/usr/bin/sh: symbolic link to bash
```

whereas on a debian box I get

```
$ file $(which sh)
/usr/bin/sh: symbolic link to dash
```

----------------------

Idea for a future release:

Replace the cram tests prompt with `bash -c '<prompt goes here>'`

**Comment by raphael-proust** [comment]
TL;DR: for opam-repo maintainers:
- The tests fail on some plateformes but it's either (a) known failures of deps build or (b) cram test failures due to different shells being the default on different plateforms.
- I think it's ready to merge.

**Comment by mseri** [comment]
Thanks
